### ISSUE TITLE
SignalR : Java Client : state and closing

### ISSUE THREAD

**Issue by TheCamel** [aspnet/AspNetCore#16598]
Hello,
my existing java client code migrated to Core cannot stop...
seems that the state never change even if i call stop() several times.. is there a workaround ?

while (Connection.getConnectionState() != HubConnectionState.DISCONNECTED) {
			try {
					Connection.stop();
					Thread.sleep(100);
				}
			catch (InterruptedException e) {
					e.printStackTrace();
				}
		}
		System.exit(0);

**Comment by BrennanConroy** [comment]
You shouldn't need to call stop in a loop. Instead you can await the stop call. Like `Connection.stop().blockingAwait();`

**Comment by TheCamel** [comment]
why not...but with previous version , i had to do this because if i call System.exit(0); before being sure the hub is closed...then signalR let a tcp connection open !

I tested what you suggest and Connection.stop().blockingAwait(); never return !?!?

**Comment by BrennanConroy** [comment]
What version of the Java client are you using?

Could you enable logging on the [client](https://docs.microsoft.com/aspnet/core/signalr/java-client?view=aspnetcore-3.0#add-logging) and the [server](https://docs.microsoft.com/aspnet/core/signalr/diagnostics?view=aspnetcore-3.0) and share them please?

**Comment by TheCamel** [comment]
i am using java 8; build 231 (1.8.0_231-b11) with signalR 1.0.0 + gson 2.8.5 

and slf4-nop 1.7.28 because was an error....
i see the disconnect event on the hub
don't know how to get more log for you...

**Comment by TheCamel** [comment]
I don't know how to get the logs....

I can send you all my code if you want, it is simple : a vs2017 solution to build the hub and an eclipse folder to debug the "java client"; 

Start the index.html (open a hub client; copy/paste the java guid into the textbox and start "simulate client"; then start the java 'applet'. in the "client session" part : Press the 'detect plugin' button to see if ok; then the 'stop' button. this send a closing message to the java part

**Comment by TheCamel** [comment]
I end up with the following:
-build;
-start;
-send message
-disconnect 
==> my close function works and state is changing

but when I send the close from the server , then it locks somewhere ?

and if I just run a timer, then it close without any trouble

Please, any comment would be nice



hubConnection.on("stopXSCP", ()-> {
						Globals.updateLastMessageReceive();
						Close(hubConnection);
				});

public static void Close(HubConnection hub)
	{
	
		while (hub.getConnectionState() != HubConnectionState.DISCONNECTED) {
			try {
				hub.stop();
					Thread.sleep(100);
				}
			catch (InterruptedException e) {
					e.printStackTrace();
				}
		}
		System.exit(0);
	}

**Comment by BrennanConroy** [comment]
I see, so you're blocking the `on` callback with a call to `stop`.

That might not work in the Java client right now. I would suggest either launching a task and running `stop` in that, or kicking off something else that runs `stop` and unblock the `on` callback as soon as possible.

Marking as investigate for the team to take a look at if this is working in the Java client.

**Comment by TheCamel** [comment]
Yes, I would say that the Stop is blocked because in the On callback... :-)
I have made a change with a watching thread to surround the problem. 

Have you seems my request about the cookies ? it is really blocking me in the migration  #16847
bye

**Comment by anurse** [comment]
Let's take a look and see if calling `stop` in an `on` callback is working.

> I can send you all my code if you want

Please do post a runnable sample @TheCamel . Even if it's "just a few projects" it's essential that we see the full context in order to understand what's going on.

**Comment by TheCamel** [comment]
have added you to acces a private repo, 

srlo.sln to create a hub

srlo.plugin is the eclipse/maven project

tell me if any trouble

**Comment by anurse** [comment]
Please ensure @BrennanConroy has access, he'll be doing a quick investigation into this.

**Comment by BrennanConroy** [comment]
> Let's take a look and see if calling `stop` in an `on` callback is working.

Verified that calling `connection.stop().blockingAwait();` will not work.

@TheCamel you can remove our access to your repo.

**Comment by anurse** [comment]
Ok, we can look resolving this or at least providing better error messaging for 5.0. For now @TheCamel you'll have to do your `stop` calls outside of `on` callbacks (even if you just fire off a new thread to do it). This is a common area we can get some deadlocks, we've seen similar behavior in .NET, so we'll take a look at finding the right solution.

**Comment by benrebak** [comment]
Is there any update on this? I get this problem when I run the following code:

```
class SignalRExample {
    public static void main(String[] args) {
        HubConnection hubConnection = HubConnectionBuilder.create("placeholder-url").build();
        hubConnection.on("Delivery", SignalRConnector::ReceiveDelivery, String.class);
        hubConnection.start().blockingAwait();

        try (Scanner reader = new Scanner(System.in)) {
            String input = "";
            while (!input.equals("leave")) {
                input = reader.nextLine();
            }
        }

        hubConnection.stop().blockingAwait();
    }

    public static void ReceiveDocumentDelivery(String delivery) {
        // ...
    }
}
```

If I run this and type leave into the console, the program doesn't stop.

**Comment by BrennanConroy** [comment]
This issue is about calling `stop` inside one of your `.on` callbacks. Looks like you're having a different problem, please file an issue with details about the problem you're experiencing.
